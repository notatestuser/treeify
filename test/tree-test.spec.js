const { asTree, asLines } = require('../treeify')

// the `asTree` expectations are prefixed with a new line so that it's more readable in jest snapshots

describe('tree-test', () => {
  describe('A tree created from an empty object when returned as a whole tree', () => {
    const subject = {}
    describe('with values hidden', () => {
      it('is an empty string', () => {
        expect(asTree(subject, false)).toBe('')
      })
    })
    describe('with values shown', () => {
      it('is an empty string', () => {
        expect(asTree(subject, true)).toBe('')
      })
    })
  })

  describe('A tree created from an object with empty prototype when returned as a whole tree', () => {
    const subject = Object.create(null)
    subject.foo = 'bar'
    describe('with values hidden', () => {
      it('produces correct output', () => {
        expect('\n' + asTree(subject, false)).toMatchSnapshot()
      })
    })
    describe('with values shown', () => {
      it('produces correct output', () => {
        expect('\n' + asTree(subject, true)).toMatchSnapshot()
      })
    })
  })

  describe('A tree created from a single-level object', () => {
    const subject = {
      apples: 'gala',
      oranges: 'mandarin',
    }
    describe('when returned line-by-line', () => {
      describe('with values hidden', () => {
        it('runs the provided callback', () => {
          const callback = jest.fn()
          asLines(subject, false, callback)
          expect(callback.mock.calls).toMatchSnapshot('calls')
        })
      })
      describe('with values shown', () => {
        it('runs the provided callback', () => {
          const callback = jest.fn()
          asLines(subject, true, callback)
          expect(callback.mock.calls).toMatchSnapshot('calls')
        })
      })
    })
    describe('when returned as a whole tree', () => {
      describe('with values hidden', () => {
        it('produces correct output', () => {
          expect('\n' + asTree(subject, false)).toMatchSnapshot()
        })
      })
      describe('with values shown', () => {
        it('produces correct output', () => {
          expect('\n' + asTree(subject, true)).toMatchSnapshot()
        })
      })
    })
  })

  describe('A tree created from a multi-level object', () => {
    const subject = {
      oranges: {
        mandarin: {
          clementine: null,
          tangerine: 'so cheap and juicy!',
        },
      },
      apples: {
        gala: null,
        'pink lady': null,
      },
    }
    const customSerializer = (v, k) => (k === 'gala' ? null : `[${k} â€“ ${v}]`)
    describe('when returned line-by-line', () => {
      describe('with values hidden', () => {
        it('runs the provided callback', () => {
          const callback = jest.fn()
          asLines(subject, false, callback)
          expect(callback.mock.calls).toMatchSnapshot('calls')
        })
      })
      describe('with values shown', () => {
        it('runs the provided callback', () => {
          const callback = jest.fn()
          asLines(subject, true, callback)
          expect(callback.mock.calls).toMatchSnapshot('calls')
        })
      })
      describe('with values shown using a custom serializer', () => {
        it('runs the provided callback', () => {
          const callback = jest.fn()
          asLines(subject, customSerializer, callback)
          expect(callback.mock.calls).toMatchSnapshot('calls')
        })
      })
    })
    describe('when returned as a whole tree', () => {
      describe('with values hidden', () => {
        it('produces correct output', () => {
          expect('\n' + asTree(subject)).toMatchSnapshot()
        })
      })
      describe('with values shown', () => {
        it('produces correct output', () => {
          expect('\n' + asTree(subject, true)).toMatchSnapshot()
        })
      })
      describe('with values shown using a custom serializer', () => {
        it('produces correct output', () => {
          expect('\n' + asTree(subject, customSerializer)).toMatchSnapshot()
        })
      })
    })
  })

  describe('A tree created from an object with not so circular references', () => {
    const subject = (() => {
      var obj = { one: 'one', two: { four: 'four' } }
      obj['three'] = obj.two
      return obj
    })()
    describe('when returned line-by-line', () => {
      describe('with values shown', () => {
        it('runs the provided callback', () => {
          const callback = jest.fn()
          asLines(subject, true, callback)
          expect(callback.mock.calls).toMatchSnapshot('calls')
        })
      })
    })
  })

  describe('A tree created from an object with circular references', () => {
    const subject = (() => {
      var obj = { one: 'one', two: 'two' }
      obj['three'] = obj
      return obj
    })()
    describe('when returned line-by-line', () => {
      describe('with values shown', () => {
        it('runs the provided callback', () => {
          const callback = jest.fn()
          asLines(subject, true, callback)
          expect(callback.mock.calls).toMatchSnapshot('calls')
        })
      })
    })
  })

  describe('A tree created from an object containing various types', () => {
    const subject = {
      array: ['one', 'two'],
      numeric: 42,
      decimal: 42.24,
      bool: false,
      nil: null,
      undef: undefined,
      date: new Date(2018, 0, 1),
    }
    describe('when returned line-by-line', () => {
      describe('with values shown', () => {
        it('runs the provided callback', () => {
          const callback = jest.fn()
          asLines(subject, true, callback)
          expect(callback.mock.calls).toMatchSnapshot('calls')
        })
      })
    })
  })

  describe('A tree created from an object with prototyped functions', () => {
    const subject = (() => {
      var func = function() {
        this.Friendly = 'stuff'
      }
      func.prototype.Nasty = function() {}
      return new func()
    })()
    describe('when returned as a whole tree', () => {
      describe('with values shown', () => {
        it('produces correct output', () => {
          expect('\n' + asTree(subject, true)).toMatchSnapshot()
        })
      })
    })
  })

  describe('A tree with functions', () => {
    const subject = {
      func: function() {},
      Friendly: 'stuff',
      Another: 'stuff',
    }
    describe('when returned line-by-line with values shown, but functions hidden', () => {
      it('runs the provided callback', () => {
        const callback = jest.fn()
        asLines(subject, false, true, callback)
        expect(callback.mock.calls).toMatchSnapshot('calls')
      })
    })
    describe('when returned as a whole tree with values shown, but functions hidden', () => {
      it('produces correct output', () => {
        expect('\n' + asTree(subject, true, true)).toMatchSnapshot()
      })
    })
  })
})
