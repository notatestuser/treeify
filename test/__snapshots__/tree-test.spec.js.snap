// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tree-test A tree created from a multi-level object when returned as a whole tree with values hidden produces correct output 1`] = `
"
├─ oranges
│  └─ mandarin
│     ├─ clementine
│     └─ tangerine
└─ apples
   ├─ gala
   └─ pink lady
"
`;

exports[`tree-test A tree created from a multi-level object when returned as a whole tree with values shown produces correct output 1`] = `
"
├─ oranges
│  └─ mandarin
│     ├─ clementine
│     └─ tangerine: so cheap and juicy!
└─ apples
   ├─ gala
   └─ pink lady
"
`;

exports[`tree-test A tree created from a multi-level object when returned as a whole tree with values shown using a custom serializer produces correct output 1`] = `
"
├─ oranges
│  └─ mandarin
│     ├─ clementine: [clementine – null]
│     └─ tangerine: [tangerine – so cheap and juicy!]
└─ apples
   ├─ gala
   └─ pink lady: [pink lady – null]
"
`;

exports[`tree-test A tree created from a multi-level object when returned line-by-line with values hidden runs the provided callback: calls 1`] = `
Array [
  Array [
    "├─ oranges",
  ],
  Array [
    "│  └─ mandarin",
  ],
  Array [
    "│     ├─ clementine",
  ],
  Array [
    "│     └─ tangerine",
  ],
  Array [
    "└─ apples",
  ],
  Array [
    "   ├─ gala",
  ],
  Array [
    "   └─ pink lady",
  ],
]
`;

exports[`tree-test A tree created from a multi-level object when returned line-by-line with values shown runs the provided callback: calls 1`] = `
Array [
  Array [
    "├─ oranges",
  ],
  Array [
    "│  └─ mandarin",
  ],
  Array [
    "│     ├─ clementine",
  ],
  Array [
    "│     └─ tangerine: so cheap and juicy!",
  ],
  Array [
    "└─ apples",
  ],
  Array [
    "   ├─ gala",
  ],
  Array [
    "   └─ pink lady",
  ],
]
`;

exports[`tree-test A tree created from a multi-level object when returned line-by-line with values shown using a custom serializer runs the provided callback: calls 1`] = `
Array [
  Array [
    "├─ oranges",
  ],
  Array [
    "│  └─ mandarin",
  ],
  Array [
    "│     ├─ clementine: [clementine – null]",
  ],
  Array [
    "│     └─ tangerine: [tangerine – so cheap and juicy!]",
  ],
  Array [
    "└─ apples",
  ],
  Array [
    "   ├─ gala",
  ],
  Array [
    "   └─ pink lady: [pink lady – null]",
  ],
]
`;

exports[`tree-test A tree created from a single-level object when returned as a whole tree with values hidden produces correct output 1`] = `
"
├─ apples
└─ oranges
"
`;

exports[`tree-test A tree created from a single-level object when returned as a whole tree with values shown produces correct output 1`] = `
"
├─ apples: gala
└─ oranges: mandarin
"
`;

exports[`tree-test A tree created from a single-level object when returned line-by-line with values hidden runs the provided callback: calls 1`] = `
Array [
  Array [
    "├─ apples",
  ],
  Array [
    "└─ oranges",
  ],
]
`;

exports[`tree-test A tree created from a single-level object when returned line-by-line with values shown runs the provided callback: calls 1`] = `
Array [
  Array [
    "├─ apples: gala",
  ],
  Array [
    "└─ oranges: mandarin",
  ],
]
`;

exports[`tree-test A tree created from an object containing various types when returned line-by-line with values shown runs the provided callback: calls 1`] = `
Array [
  Array [
    "├─ array",
  ],
  Array [
    "│  ├─ 0: one",
  ],
  Array [
    "│  └─ 1: two",
  ],
  Array [
    "├─ numeric: 42",
  ],
  Array [
    "├─ decimal: 42.24",
  ],
  Array [
    "├─ bool: false",
  ],
  Array [
    "├─ nil",
  ],
  Array [
    "├─ undef: undefined",
  ],
  Array [
    "└─ date: 2018-01-01T00:00:00.000Z",
  ],
]
`;

exports[`tree-test A tree created from an object with circular references when returned line-by-line with values shown runs the provided callback: calls 1`] = `
Array [
  Array [
    "├─ one: one",
  ],
  Array [
    "├─ two: two",
  ],
  Array [
    "└─ three (circular ref.)",
  ],
]
`;

exports[`tree-test A tree created from an object with empty prototype when returned as a whole tree with values hidden produces correct output 1`] = `
"
└─ foo
"
`;

exports[`tree-test A tree created from an object with empty prototype when returned as a whole tree with values shown produces correct output 1`] = `
"
└─ foo: bar
"
`;

exports[`tree-test A tree created from an object with not so circular references when returned line-by-line with values shown runs the provided callback: calls 1`] = `
Array [
  Array [
    "├─ one: one",
  ],
  Array [
    "├─ two",
  ],
  Array [
    "│  └─ four: four",
  ],
  Array [
    "└─ three",
  ],
  Array [
    "   └─ four: four",
  ],
]
`;

exports[`tree-test A tree created from an object with prototyped functions when returned as a whole tree with values shown produces correct output 1`] = `
"
└─ Friendly: stuff
"
`;

exports[`tree-test A tree with functions when returned as a whole tree with values shown, but functions hidden produces correct output 1`] = `
"
├─ Friendly: stuff
└─ Another: stuff
"
`;

exports[`tree-test A tree with functions when returned line-by-line with values shown, but functions hidden runs the provided callback: calls 1`] = `
Array [
  Array [
    "├─ Friendly",
  ],
  Array [
    "└─ Another",
  ],
]
`;
